CSE 354 - Project: Sentiment Serenade

By Andrew Sun, Isaac Jon, Kevin Mai

Install transformers first

In [ ]:
!pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.0/7.0 MB 46.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 224.5/224.5 kB 18.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 85.9 MB/s eta 0:00:00


Imports

In [ ]:
from collections import Counter
from typing import List, Dict, Tuple, Any
import json
import os
import zipfile
import argparse
import random
import copy
import re
import nltk

from transformers import AutoTokenizer, AutoModelForSequenceClassification
from transformers import AdamW

import torch
import torchtext
from torchtext.data.utils import get_tokenizer
from torchtext.vocab import build_vocab_from_iterator
import pandas as pd
from sklearn.model_selection import train_test_split
from torch.utils.data import Dataset, TensorDataset, DataLoader
from torch.nn.utils.rnn import pad_sequence
from torch.nn import Transformer
from tqdm import tqdm
import numpy as np
from sklearn.metrics import precision_score, recall_score, f1_score
torch.manual_seed(42)
np.random.seed(42)

ModuleNotFoundError: ignored

Mount Drive First

In [ ]:
from google.colab import drive
drive.mount('/content/drive',force_remount=True)

# TODO(students): start
#%cd "drive/MyDrive/--------"
%cd "/content/drive/MyDrive/Project"
# TODO(students): end

Mounted at /content/drive
/content/drive/MyDrive/Project


List current directory structure

In [ ]:
import locale
def getpreferredencoding(do_setlocale = True):
    return "UTF-8"
locale.getpreferredencoding = getpreferredencoding
!sudo apt install tree
!tree

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following NEW packages will be installed:
  tree
0 upgraded, 1 newly installed, 0 to remove and 24 not upgraded.
Need to get 43.0 kB of archives.
After this operation, 115 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu focal/universe amd64 tree amd64 1.8.0-1 [43.0 kB]
Fetched 43.0 kB in 1s (49.9 kB/s)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 1.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Selecting previously unselected package tree.
(Reading database ... 122518 files and directories currently install

CONSTANTS for this project

In [ ]:
#DO NOT CHANGE THE CONSTANTS
BATCH_SIZE = 16
EPOCHS = 3
RANDOM_TEST_PATH = "data/random_test.csv"
FIXED_TEST_PATH = "data/fixed_test.csv"
TRAIN_PATH = "data/train.csv"
VAL_PATH = "data/dev.csv"
ROBERTATINY_SAVE_PATH = "models/RobertaTiny"
DISTILBERT_SAVE_PATH = "models/DistilBERT"

In [ ]:
def load_dataset(path):
  dataset = pd.read_csv(path)
  return dataset

In [ ]:
train_data = load_dataset(TRAIN_PATH)
random_test_data = load_dataset(RANDOM_TEST_PATH)
fixed_test_data = load_dataset(FIXED_TEST_PATH)
val_data = load_dataset(VAL_PATH)

Method 1: Training a network from scratch

In [ ]:
class RobertaTiny():

  def __init__(self, model_name='arampacha/roberta-tiny', num_classes=3):
    # TODO(students): start
    self.tokenizer = AutoTokenizer.from_pretrained(model_name)
    self.model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_classes)
    # TODO(students): end
  def get_tokenizer_and_model(self):
    return self.model, self.tokenizer

In [ ]:
class DatasetLoader(Dataset):

  def __init__(self, data, tokenizer):
    self.data = data
    self.tokenizer = tokenizer

  def tokenize_data(self):
    print("Processing data..")
    tokens = []
    labels = []
    label_dict = {'Positive': 2, 'Neutral': 1, 'Negative': 0}

    document_list = self.data['DOCUMENT'].to_list()
    label_list = self.data['TRUE_SENTIMENT'].to_list()

    for (document, label) in tqdm(zip(document_list, label_list), total=len(document_list)):
      # TODO(students): start
      encoded_dictionary = self.tokenizer.encode_plus(document, truncation = True, max_length=256, return_tensors = "pt")
      document_tokens = encoded_dictionary["input_ids"][0]
      tokens.append(document_tokens)
      labels.append(label_dict[label])
      # TODO(students): end
    
    tokens = pad_sequence(tokens, batch_first=True)
    labels = torch.tensor(labels)
    dataset = TensorDataset(tokens, labels)
    return dataset

  def get_data_loaders(self, batch_size=32, shuffle=True):
    processed_dataset = self.tokenize_data()

    data_loader = DataLoader(
        processed_dataset,
        shuffle=shuffle,
        batch_size=batch_size
    )

    return data_loader

In [ ]:
class Trainer():

  def __init__(self, options):
    self.device = options['device']
    self.train_data = options['train_data']
    self.val_data = options['val_data']
    self.batch_size = options['batch_size']
    self.epochs = options['epochs']
    self.save_path = options['save_path']
    self.training_type = options['training_type']
    transformer = RobertaTiny()
    self.model, self.tokenizer = transformer.get_tokenizer_and_model()
    self.model.to(self.device)

  def get_performance_metrics(self, preds, labels):
    pred_flat = np.argmax(preds, axis=1).flatten()
    labels_flat = labels.flatten()
    precision = precision_score(labels_flat, pred_flat, zero_division=0, average='weighted')
    recall = recall_score(labels_flat, pred_flat, zero_division=0, average='weighted')
    f1 = f1_score(labels_flat, pred_flat, zero_division=0, average='weighted')
    return precision, recall, f1

  def set_training_parameters(self):
    # TODO(students): start
    if self.training_type == 'frozen_embeddings':
      for layer in self.model.roberta.encoder.layer:
        layer.requires_grad = False
    elif self.training_type == "top_1_training":
      for i in range(len(self.model.roberta.encoder.layer) - 1):
        layer = self.model.roberta.encoder.layer[i]
        layer.requires_grad = False
    elif self.training_type == "top_2_training":
      for i in range(len(self.model.roberta.encoder.layer) - 2):
        layer = self.model.roberta.encoder.layer[i]
        layer.requires_grad = False
    elif self.training_type == "all_training":
      pass

    # TODO(students): end

  def train(self, data_loader, optimizer):
    self.model.train()
    total_recall = 0
    total_precision = 0
    total_f1 = 0
    total_loss = 0

    for batch_idx, (reviews, labels) in enumerate(tqdm(data_loader)):
      self.model.zero_grad()
      # TODO(students): start

      reviews = reviews.to(self.device)
      labels = labels.to(self.device)

      outputs = self.model(reviews, labels=labels)

      batch_loss = outputs.loss
      batch_logits = outputs.logits

      optimizer.zero_grad()
      batch_loss.backward()
      optimizer.step()
      
      # get precision, recall, f1.
      (batch_precision, batch_recall, batch_f1) = self.get_performance_metrics(batch_logits.detach().cpu().numpy(), labels.detach().cpu().numpy())
      total_precision += batch_precision
      total_recall += batch_recall
      total_f1 += batch_f1
      total_loss += batch_loss

      # TODO(students): end

    precision = total_precision/len(data_loader)
    recall = total_recall/len(data_loader)
    f1 = total_f1/len(data_loader)
    loss = total_loss/len(data_loader)

    return precision, recall, f1, loss

  def eval(self, data_loader):
    self.model.eval()
    total_recall = 0
    total_precision = 0
    total_f1 = 0
    total_loss = 0

    with torch.no_grad():
      for (reviews, labels) in tqdm(data_loader):
        # TODO(students): start

        reviews = reviews.to(self.device)
        labels = labels.to(self.device)
        
        outputs = self.model(reviews, labels=labels)

        batch_loss = outputs.loss
        batch_logits = outputs.logits

        # get precision, recall, f1.
        (batch_precision, batch_recall, batch_f1) = self.get_performance_metrics(batch_logits.detach().cpu().numpy(), labels.detach().cpu().numpy())
        total_precision += batch_precision
        total_recall += batch_recall
        total_f1 += batch_f1
        total_loss += batch_loss

        # TODO(students): end
    
    precision = total_precision/len(data_loader)
    recall = total_recall/len(data_loader)
    f1 = total_f1/len(data_loader)
    loss = total_loss/len(data_loader)

    return precision, recall, f1, loss

  def save_transformer(self):
    self.model.save_pretrained(self.save_path)
    self.tokenizer.save_pretrained(self.save_path)

  def execute(self):
    last_best = 0
    train_dataset = DatasetLoader(self.train_data, self.tokenizer)
    train_data_loader = train_dataset.get_data_loaders(self.batch_size)
    val_dataset = DatasetLoader(self.val_data, self.tokenizer)
    val_data_loader = val_dataset.get_data_loaders(self.batch_size)
    optimizer = torch.optim.AdamW(self.model.parameters(), lr = 3e-5, eps = 1e-8)
    self.set_training_parameters()
    for epoch_i in range(0, self.epochs):
      train_precision, train_recall, train_f1, train_loss = self.train(train_data_loader, optimizer)
      print(f'Epoch {epoch_i + 1}: train_loss: {train_loss:.4f} train_precision: {train_precision:.4f} train_recall: {train_recall:.4f} train_f1: {train_f1:.4f}')
      val_precision, val_recall, val_f1, val_loss = self.eval(val_data_loader)
      print(f'Epoch {epoch_i + 1}: val_loss: {val_loss:.4f} val_precision: {val_precision:.4f} val_recall: {val_recall:.4f} val_f1: {val_f1:.4f}')

      if val_f1 > last_best:
        print("Saving model..")
        self.save_transformer()
        last_best = val_f1
        print("Model saved.")

In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['train_data'] = train_data
options['val_data'] = val_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_frozen_embeddings'
options['epochs'] = EPOCHS
options['training_type'] = 'frozen_embeddings'
trainer = Trainer(options)
trainer.execute()

Some weights of the model checkpoint at arampacha/roberta-tiny were not used when initializing RobertaForSequenceClassification: ['lm_head.bias', 'lm_head.dense.weight', 'lm_head.layer_norm.weight', 'lm_head.dense.bias', 'lm_head.layer_norm.bias']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at arampacha/roberta-tiny and are newly initialized: ['classifier.dense.bias', 'classifier.out_proj.weight', 'classifier.out_proj.bias', 'classifier.dense.we

Processing data..


100%|██████████| 3355/3355 [00:12<00:00, 277.92it/s]


Processing data..


100%|██████████| 210/210 [00:14<00:00, 14.96it/s]


Epoch 1: train_loss: 0.9530 train_precision: 0.3255 train_recall: 0.5202 train_f1: 0.3761


100%|██████████| 37/37 [00:00<00:00, 54.66it/s]


Epoch 1: val_loss: 0.9480 val_precision: 0.2874 val_recall: 0.5135 val_f1: 0.3646
Saving model..
Model saved.


100%|██████████| 210/210 [00:11<00:00, 18.15it/s]


Epoch 2: train_loss: 0.9451 train_precision: 0.2922 train_recall: 0.5212 train_f1: 0.3691


100%|██████████| 37/37 [00:00<00:00, 55.54it/s]


Epoch 2: val_loss: 0.9506 val_precision: 0.2872 val_recall: 0.5135 val_f1: 0.3646


100%|██████████| 210/210 [00:11<00:00, 17.92it/s]


Epoch 3: train_loss: 0.9329 train_precision: 0.4133 train_recall: 0.5280 train_f1: 0.4345


100%|██████████| 37/37 [00:00<00:00, 55.11it/s]


Epoch 3: val_loss: 0.9312 val_precision: 0.4796 val_recall: 0.5473 val_f1: 0.4355
Saving model..
Model saved.


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['train_data'] = train_data
options['val_data'] = val_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_top_1_training'
options['epochs'] = EPOCHS
options['training_type'] = 'top_1_training'
trainer = Trainer(options)
trainer.execute()

Some weights of the model checkpoint at arampacha/roberta-tiny were not used when initializing RobertaForSequenceClassification: ['lm_head.bias', 'lm_head.dense.weight', 'lm_head.layer_norm.weight', 'lm_head.dense.bias', 'lm_head.layer_norm.bias']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at arampacha/roberta-tiny and are newly initialized: ['classifier.dense.bias', 'classifier.out_proj.weight', 'classifier.out_proj.bias', 'classifier.dense.we

Processing data..


100%|██████████| 3355/3355 [00:06<00:00, 491.37it/s]


Processing data..


100%|██████████| 210/210 [00:11<00:00, 17.57it/s]


Epoch 1: train_loss: 0.9552 train_precision: 0.3029 train_recall: 0.5170 train_f1: 0.3723


100%|██████████| 37/37 [00:00<00:00, 53.96it/s]


Epoch 1: val_loss: 0.9289 val_precision: 0.3174 val_recall: 0.5372 val_f1: 0.3904
Saving model..
Model saved.


100%|██████████| 210/210 [00:11<00:00, 17.81it/s]


Epoch 2: train_loss: 0.9440 train_precision: 0.3022 train_recall: 0.5225 train_f1: 0.3731


100%|██████████| 37/37 [00:00<00:00, 52.10it/s]


Epoch 2: val_loss: 0.9396 val_precision: 0.2891 val_recall: 0.5135 val_f1: 0.3648


100%|██████████| 210/210 [00:11<00:00, 17.98it/s]


Epoch 3: train_loss: 0.9351 train_precision: 0.4134 train_recall: 0.5279 train_f1: 0.4380


100%|██████████| 37/37 [00:00<00:00, 50.23it/s]


Epoch 3: val_loss: 0.9371 val_precision: 0.4560 val_recall: 0.4848 val_f1: 0.4567
Saving model..
Model saved.


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['train_data'] = train_data
options['val_data'] = val_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_top_2_training'
options['epochs'] = EPOCHS
options['training_type'] = 'top_2_training'
trainer = Trainer(options)
trainer.execute()

Some weights of the model checkpoint at arampacha/roberta-tiny were not used when initializing RobertaForSequenceClassification: ['lm_head.bias', 'lm_head.dense.weight', 'lm_head.layer_norm.weight', 'lm_head.dense.bias', 'lm_head.layer_norm.bias']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at arampacha/roberta-tiny and are newly initialized: ['classifier.dense.bias', 'classifier.out_proj.weight', 'classifier.out_proj.bias', 'classifier.dense.we

Processing data..


100%|██████████| 3355/3355 [00:05<00:00, 565.35it/s]


Processing data..


100%|██████████| 210/210 [00:11<00:00, 18.16it/s]


Epoch 1: train_loss: 0.9542 train_precision: 0.3071 train_recall: 0.5209 train_f1: 0.3737


100%|██████████| 37/37 [00:00<00:00, 56.15it/s]


Epoch 1: val_loss: 0.9521 val_precision: 0.2896 val_recall: 0.5253 val_f1: 0.3694
Saving model..
Model saved.


100%|██████████| 210/210 [00:11<00:00, 17.98it/s]


Epoch 2: train_loss: 0.9438 train_precision: 0.2992 train_recall: 0.5240 train_f1: 0.3727


100%|██████████| 37/37 [00:00<00:00, 56.25it/s]


Epoch 2: val_loss: 0.9468 val_precision: 0.2879 val_recall: 0.5253 val_f1: 0.3686


100%|██████████| 210/210 [00:11<00:00, 18.22it/s]


Epoch 3: train_loss: 0.9352 train_precision: 0.4123 train_recall: 0.5287 train_f1: 0.4280


100%|██████████| 37/37 [00:00<00:00, 50.84it/s]


Epoch 3: val_loss: 0.9334 val_precision: 0.5170 val_recall: 0.5456 val_f1: 0.4678
Saving model..
Model saved.


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['train_data'] = train_data
options['val_data'] = val_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_all_training'
options['epochs'] = EPOCHS
options['training_type'] = 'all_training'
trainer = Trainer(options)
trainer.execute()

Some weights of the model checkpoint at arampacha/roberta-tiny were not used when initializing RobertaForSequenceClassification: ['lm_head.bias', 'lm_head.dense.weight', 'lm_head.layer_norm.weight', 'lm_head.dense.bias', 'lm_head.layer_norm.bias']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at arampacha/roberta-tiny and are newly initialized: ['classifier.dense.bias', 'classifier.out_proj.weight', 'classifier.out_proj.bias', 'classifier.dense.we

Processing data..


100%|██████████| 3355/3355 [00:05<00:00, 574.62it/s]


Processing data..


100%|██████████| 210/210 [00:11<00:00, 18.02it/s]


Epoch 1: train_loss: 0.9522 train_precision: 0.3029 train_recall: 0.5191 train_f1: 0.3700


100%|██████████| 37/37 [00:00<00:00, 55.12it/s]


Epoch 1: val_loss: 0.9347 val_precision: 0.2953 val_recall: 0.5253 val_f1: 0.3728
Saving model..
Model saved.


100%|██████████| 210/210 [00:12<00:00, 17.41it/s]


Epoch 2: train_loss: 0.9421 train_precision: 0.3250 train_recall: 0.5236 train_f1: 0.3826


100%|██████████| 37/37 [00:00<00:00, 55.22it/s]


Epoch 2: val_loss: 0.9491 val_precision: 0.3689 val_recall: 0.5236 val_f1: 0.3870
Saving model..
Model saved.


100%|██████████| 210/210 [00:11<00:00, 18.07it/s]


Epoch 3: train_loss: 0.9319 train_precision: 0.4561 train_recall: 0.5331 train_f1: 0.4509


100%|██████████| 37/37 [00:00<00:00, 50.18it/s]


Epoch 3: val_loss: 0.9301 val_precision: 0.4837 val_recall: 0.5118 val_f1: 0.4684
Saving model..
Model saved.


Tester

In [ ]:
class Tester():

  def __init__(self, options):
    self.save_path = options['save_path']
    self.device = options['device']
    self.test_data = options['test_data']
    self.batch_size = options['batch_size']
    transformer = RobertaTiny(self.save_path)
    self.model, self.tokenizer = transformer.get_tokenizer_and_model()
    self.model.to(self.device)

  def get_performance_metrics(self, preds, labels):
    pred_flat = np.argmax(preds, axis=1).flatten()
    labels_flat = labels.flatten()
    precision = precision_score(labels_flat, pred_flat, zero_division=0, average="weighted")
    recall = recall_score(labels_flat, pred_flat, zero_division=0, average="weighted")
    f1 = f1_score(labels_flat, pred_flat, zero_division=0, average="weighted")
    return precision, recall, f1

  def test(self, data_loader):
    self.model.eval()
    total_recall = 0
    total_precision = 0
    total_f1 = 0
    total_loss = 0

    with torch.no_grad():
      for (reviews, labels) in tqdm(data_loader):
        # TODO(students): start

        reviews = reviews.to(self.device)
        labels = labels.to(self.device)

        outputs = self.model(reviews, labels=labels)

        batch_loss = outputs.loss
        batch_logits = outputs.logits

        # get precision, recall, f1.
        (batch_precision, batch_recall, batch_f1) = self.get_performance_metrics(batch_logits.detach().cpu().numpy(), labels.detach().cpu().numpy())
        total_precision += batch_precision
        total_recall += batch_recall
        total_f1 += batch_f1
        total_loss += batch_loss

        # TODO(students): end
    
    precision = total_precision/len(data_loader)
    recall = total_recall/len(data_loader)
    f1 = total_f1/len(data_loader)
    loss = total_loss/len(data_loader)

    return precision, recall, f1, loss

  def execute(self):
    test_dataset = DatasetLoader(self.test_data, self.tokenizer)
    test_data_loader = test_dataset.get_data_loaders(self.batch_size)

    test_precision, test_recall, test_f1, test_loss = self.test(test_data_loader)

    print()
    print(f'test_loss: {test_loss:.4f} test_precision: {test_precision:.4f} test_recall: {test_recall:.4f} test_f1: {test_f1:.4f}')

Testing the embeddings

With Fixed Test Data

In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = fixed_test_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_frozen_embeddings'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 52/52 [00:01<00:00, 49.51it/s]


test_loss: 1.0520 test_precision: 0.3905 test_recall: 0.4569 test_f1: 0.3440


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = fixed_test_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_top_1_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 52/52 [00:01<00:00, 50.58it/s]


test_loss: 1.0522 test_precision: 0.4078 test_recall: 0.4486 test_f1: 0.4102


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = fixed_test_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_top_2_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 52/52 [00:00<00:00, 52.48it/s]


test_loss: 1.0565 test_precision: 0.3857 test_recall: 0.4569 test_f1: 0.3747


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = fixed_test_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_all_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 52/52 [00:01<00:00, 45.57it/s]


test_loss: 1.0635 test_precision: 0.4019 test_recall: 0.4497 test_f1: 0.4010


With Random Test Data

In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = random_test_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_frozen_embeddings'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 37/37 [00:00<00:00, 44.24it/s]


test_loss: 0.9768 test_precision: 0.3941 test_recall: 0.5090 test_f1: 0.3874


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = random_test_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_top_1_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 37/37 [00:00<00:00, 49.47it/s]


test_loss: 0.9706 test_precision: 0.4381 test_recall: 0.4854 test_f1: 0.4427


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = random_test_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_top_2_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 37/37 [00:00<00:00, 41.74it/s]


test_loss: 0.9696 test_precision: 0.4347 test_recall: 0.5068 test_f1: 0.4189


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = random_test_data
options['save_path'] = ROBERTATINY_SAVE_PATH + '_all_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 37/37 [00:00<00:00, 44.26it/s]


test_loss: 0.9676 test_precision: 0.4657 test_recall: 0.5220 test_f1: 0.4704


Method 2: Fine-tuning a small sized pre-trained model (DistilBERT)

In [ ]:
class DistillBERT():

  def __init__(self, model_name='distilbert-base-uncased', num_classes=3):
    # TODO(students): start
    self.tokenizer = AutoTokenizer.from_pretrained(model_name)
    self.model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=num_classes)
    # TODO(students): end

  def get_tokenizer_and_model(self):
    return self.model, self.tokenizer

In [ ]:
class DatasetLoader(Dataset):

  def __init__(self, data, tokenizer):
    self.data = data
    self.tokenizer = tokenizer

  def tokenize_data(self):
    print("Processing data..")
    tokens = []
    labels = []
    label_dict = {'Positive': 2, 'Neutral': 1, 'Negative': 0}

    document_list = self.data['DOCUMENT'].to_list()
    label_list = self.data['TRUE_SENTIMENT'].to_list()
    # print(label_list)

    for (document, label) in tqdm(zip(document_list, label_list), total=len(document_list)):
      # TODO(students): start
      encoded_dictionary = self.tokenizer.encode_plus(document, truncation = True, return_tensors = "pt")
      document_tokens = encoded_dictionary["input_ids"][0]
      tokens.append(document_tokens)
      labels.append(label_dict[label])
      # TODO(students): end
    
    tokens = pad_sequence(tokens, batch_first=True)
    labels = torch.tensor(labels)
    # print(labels)
    dataset = TensorDataset(tokens, labels)
    return dataset

  def get_data_loaders(self, batch_size=32, shuffle=True):
    processed_dataset = self.tokenize_data()

    data_loader = DataLoader(
        processed_dataset,
        shuffle=False,
        batch_size=batch_size
    )

    return data_loader

In [ ]:
class Trainer():

  def __init__(self, options):
    self.device = options['device']
    self.train_data = options['train_data']
    self.val_data = options['val_data']
    self.batch_size = options['batch_size']
    self.epochs = options['epochs']
    self.save_path = options['save_path']
    self.training_type = options['training_type']
    transformer = DistillBERT()
    self.model, self.tokenizer = transformer.get_tokenizer_and_model()
    self.model.to(self.device)

  def get_performance_metrics(self, preds, labels):
    pred_flat = np.argmax(preds, axis=1).flatten()
    labels_flat = labels.flatten()
    precision = precision_score(labels_flat, pred_flat, zero_division=0, average='weighted')
    recall = recall_score(labels_flat, pred_flat, zero_division=0, average='weighted')
    f1 = f1_score(labels_flat, pred_flat, zero_division=0, average='weighted')
    return precision, recall, f1

  def set_training_parameters(self):
    # TODO(students): start
    if self.training_type == 'frozen_embeddings':
      for layer in self.model.distilbert.transformer.layer:
        layer.requires_grad = False
    elif self.training_type == "top_2_training":
      for i in range(len(self.model.distilbert.transformer.layer) - 2):
        layer = self.model.distilbert.transformer.layer[i]
        layer.requires_grad = False
    elif self.training_type == "top_4_training":
      for i in range(len(self.model.distilbert.transformer.layer) - 4):
        layer = self.model.distilbert.transformer.layer[i]
        layer.requires_grad = False
    elif self.training_type == "all_training":
      pass
    # TODO(students): end

  def train(self, data_loader, optimizer):
    self.model.train()
    total_recall = 0
    total_precision = 0
    total_f1 = 0
    total_loss = 0

    for batch_idx, (reviews, labels) in enumerate(tqdm(data_loader)):
      self.model.zero_grad()
      # TODO(students): start

      reviews = reviews.to(self.device)
      labels = labels.to(self.device)

      outputs = self.model(reviews, labels=labels)

      batch_loss = outputs.loss
      batch_logits = outputs.logits

      optimizer.zero_grad()
      batch_loss.backward()
      optimizer.step()
      
      # get precision, recall, f1.
      (batch_precision, batch_recall, batch_f1) = self.get_performance_metrics(batch_logits.detach().cpu().numpy(), labels.detach().cpu().numpy())
      total_precision += batch_precision
      total_recall += batch_recall
      total_f1 += batch_f1
      total_loss += batch_loss

      # TODO(students): end

    precision = total_precision/len(data_loader)
    recall = total_recall/len(data_loader)
    f1 = total_f1/len(data_loader)
    loss = total_loss/len(data_loader)

    return precision, recall, f1, loss

  def eval(self, data_loader):
    self.model.eval()
    total_recall = 0
    total_precision = 0
    total_f1 = 0
    total_loss = 0

    with torch.no_grad():
      for (reviews, labels) in tqdm(data_loader):
        # TODO(students): start

        reviews = reviews.to(self.device)
        labels = labels.to(self.device)

        print(labels)
        
        outputs = self.model(reviews, labels=labels)

        batch_loss = outputs.loss
        batch_logits = outputs.logits

        # get precision, recall, f1.
        (batch_precision, batch_recall, batch_f1) = self.get_performance_metrics(batch_logits.detach().cpu().numpy(), labels.detach().cpu().numpy())
        total_precision += batch_precision
        total_recall += batch_recall
        total_f1 += batch_f1
        total_loss += batch_loss

        # TODO(students): end
    
    precision = total_precision/len(data_loader)
    recall = total_recall/len(data_loader)
    f1 = total_f1/len(data_loader)
    loss = total_loss/len(data_loader)

    return precision, recall, f1, loss

  def save_transformer(self):
    self.model.save_pretrained(self.save_path)
    self.tokenizer.save_pretrained(self.save_path)

  def execute(self):
    last_best = 0
    train_dataset = DatasetLoader(self.train_data, self.tokenizer)
    train_data_loader = train_dataset.get_data_loaders(self.batch_size)
    val_dataset = DatasetLoader(self.val_data, self.tokenizer)
    val_data_loader = val_dataset.get_data_loaders(self.batch_size)
    optimizer = torch.optim.AdamW(self.model.parameters(), lr = 3e-5, eps = 1e-8)
    self.set_training_parameters()
    for epoch_i in range(0, self.epochs):
      train_precision, train_recall, train_f1, train_loss = self.train(train_data_loader, optimizer)
      print(f'Epoch {epoch_i + 1}: train_loss: {train_loss:.4f} train_precision: {train_precision:.4f} train_recall: {train_recall:.4f} train_f1: {train_f1:.4f}')
      val_precision, val_recall, val_f1, val_loss = self.eval(val_data_loader)
      print(f'Epoch {epoch_i + 1}: val_loss: {val_loss:.4f} val_precision: {val_precision:.4f} val_recall: {val_recall:.4f} val_f1: {val_f1:.4f}')

      if val_f1 > last_best:
        print("Saving model..")
        self.save_transformer()
        last_best = val_f1
        print("Model saved.")

In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['train_data'] = train_data
options['val_data'] = val_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_frozen_embeddings'
options['epochs'] = EPOCHS
options['training_type'] = 'frozen_embeddings'
trainer = Trainer(options)
trainer.execute()

Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertForSequenceClassification: ['vocab_transform.bias', 'vocab_projector.weight', 'vocab_layer_norm.weight', 'vocab_transform.weight', 'vocab_layer_norm.bias', 'vocab_projector.bias']
- This IS expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['pre_classifier.weight', 'classifier.weight', 'classifi

Processing data..


100%|██████████| 3355/3355 [00:05<00:00, 579.88it/s]


Processing data..


100%|██████████| 210/210 [02:34<00:00,  1.36it/s]


Epoch 1: train_loss: 0.9285 train_precision: 0.4255 train_recall: 0.5254 train_f1: 0.4353


100%|██████████| 37/37 [00:09<00:00,  4.03it/s]


Epoch 1: val_loss: 0.8641 val_precision: 0.5703 val_recall: 0.6014 val_f1: 0.5735
Saving model..
Model saved.


100%|██████████| 210/210 [02:37<00:00,  1.33it/s]


Epoch 2: train_loss: 0.8502 train_precision: 0.5601 train_recall: 0.5871 train_f1: 0.5467


100%|██████████| 37/37 [00:09<00:00,  3.98it/s]


Epoch 2: val_loss: 0.8501 val_precision: 0.5519 val_recall: 0.5811 val_f1: 0.5542


100%|██████████| 210/210 [02:38<00:00,  1.33it/s]


Epoch 3: train_loss: 0.7417 train_precision: 0.6437 train_recall: 0.6496 train_f1: 0.6249


100%|██████████| 37/37 [00:09<00:00,  3.96it/s]


Epoch 3: val_loss: 0.8520 val_precision: 0.5801 val_recall: 0.6115 val_f1: 0.5827
Saving model..
Model saved.


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['train_data'] = train_data
options['val_data'] = val_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_top_2_training'
options['epochs'] = EPOCHS
options['training_type'] = 'top_2_training'
trainer = Trainer(options)
trainer.execute()

Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertForSequenceClassification: ['vocab_transform.bias', 'vocab_projector.weight', 'vocab_layer_norm.weight', 'vocab_transform.weight', 'vocab_layer_norm.bias', 'vocab_projector.bias']
- This IS expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['pre_classifier.weight', 'classifier.weight', 'classifi

Processing data..


100%|██████████| 3355/3355 [00:06<00:00, 507.24it/s]


Processing data..


100%|██████████| 210/210 [02:38<00:00,  1.33it/s]


Epoch 1: train_loss: 0.9241 train_precision: 0.4397 train_recall: 0.5316 train_f1: 0.4448


100%|██████████| 37/37 [00:09<00:00,  3.98it/s]


Epoch 1: val_loss: 0.8436 val_precision: 0.5991 val_recall: 0.6014 val_f1: 0.5815
Saving model..
Model saved.


100%|██████████| 210/210 [02:38<00:00,  1.33it/s]


Epoch 2: train_loss: 0.8602 train_precision: 0.5496 train_recall: 0.5740 train_f1: 0.5317


100%|██████████| 37/37 [00:09<00:00,  3.96it/s]


Epoch 2: val_loss: 0.8482 val_precision: 0.5447 val_recall: 0.5997 val_f1: 0.5536


100%|██████████| 210/210 [02:38<00:00,  1.33it/s]


Epoch 3: train_loss: 0.7405 train_precision: 0.6535 train_recall: 0.6531 train_f1: 0.6307


100%|██████████| 37/37 [00:09<00:00,  3.96it/s]

Epoch 3: val_loss: 0.8482 val_precision: 0.5704 val_recall: 0.5709 val_f1: 0.5562


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['train_data'] = train_data
options['val_data'] = val_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_top_4_training'
options['epochs'] = EPOCHS
options['training_type'] = 'top_4_training'
trainer = Trainer(options)
trainer.execute()

Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertForSequenceClassification: ['vocab_transform.bias', 'vocab_projector.weight', 'vocab_layer_norm.weight', 'vocab_transform.weight', 'vocab_layer_norm.bias', 'vocab_projector.bias']
- This IS expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['pre_classifier.weight', 'classifier.weight', 'classifi

Processing data..


100%|██████████| 3355/3355 [00:04<00:00, 673.41it/s]


Processing data..


100%|██████████| 210/210 [02:38<00:00,  1.32it/s]


Epoch 1: train_loss: 0.9471 train_precision: 0.3651 train_recall: 0.5169 train_f1: 0.4030


100%|██████████| 37/37 [00:09<00:00,  3.98it/s]


Epoch 1: val_loss: 0.8931 val_precision: 0.5428 val_recall: 0.5828 val_f1: 0.5068
Saving model..
Model saved.


100%|██████████| 210/210 [02:38<00:00,  1.33it/s]


Epoch 2: train_loss: 0.8870 train_precision: 0.5257 train_recall: 0.5634 train_f1: 0.5129


100%|██████████| 37/37 [00:09<00:00,  3.98it/s]


Epoch 2: val_loss: 0.8342 val_precision: 0.5439 val_recall: 0.5963 val_f1: 0.5385
Saving model..
Model saved.


100%|██████████| 210/210 [02:38<00:00,  1.33it/s]


Epoch 3: train_loss: 0.8058 train_precision: 0.5977 train_recall: 0.6248 train_f1: 0.5898


100%|██████████| 37/37 [00:09<00:00,  3.97it/s]

Epoch 3: val_loss: 0.8716 val_precision: 0.5570 val_recall: 0.5203 val_f1: 0.4941


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['train_data'] = train_data
options['val_data'] = val_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_all_training'
options['epochs'] = EPOCHS
options['training_type'] = 'all_training'
trainer = Trainer(options)
trainer.execute()

Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertForSequenceClassification: ['vocab_transform.bias', 'vocab_projector.weight', 'vocab_layer_norm.weight', 'vocab_transform.weight', 'vocab_layer_norm.bias', 'vocab_projector.bias']
- This IS expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of DistilBertForSequenceClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['pre_classifier.weight', 'classifier.weight', 'classifi

Processing data..


100%|██████████| 3355/3355 [00:05<00:00, 663.24it/s]


Processing data..


100%|██████████| 210/210 [02:38<00:00,  1.33it/s]


Epoch 1: train_loss: 0.9443 train_precision: 0.3944 train_recall: 0.5194 train_f1: 0.4164


100%|██████████| 37/37 [00:09<00:00,  3.98it/s]


Epoch 1: val_loss: 0.9246 val_precision: 0.5865 val_recall: 0.5051 val_f1: 0.4519
Saving model..
Model saved.


100%|██████████| 210/210 [02:37<00:00,  1.33it/s]


Epoch 2: train_loss: 0.8927 train_precision: 0.5107 train_recall: 0.5548 train_f1: 0.5026


100%|██████████| 37/37 [00:09<00:00,  4.01it/s]


Epoch 2: val_loss: 0.8552 val_precision: 0.5670 val_recall: 0.6064 val_f1: 0.5691
Saving model..
Model saved.


100%|██████████| 210/210 [02:38<00:00,  1.33it/s]


Epoch 3: train_loss: 0.8012 train_precision: 0.5925 train_recall: 0.6160 train_f1: 0.5805


100%|██████████| 37/37 [00:09<00:00,  3.98it/s]

Epoch 3: val_loss: 0.8413 val_precision: 0.5783 val_recall: 0.5912 val_f1: 0.5640


Tester

In [ ]:
class Tester():

  def __init__(self, options):
    self.save_path = options['save_path']
    self.device = options['device']
    self.test_data = options['test_data']
    self.batch_size = options['batch_size']
    transformer = DistillBERT(self.save_path)
    self.model, self.tokenizer = transformer.get_tokenizer_and_model()
    self.model.to(self.device)
    # self.preds = []
    # self.labels = []

  def get_performance_metrics(self, preds, labels):
    pred_flat = np.argmax(preds, axis=1).flatten()
    # for num in pred_flat:
    #   self.preds.append(num)
    labels_flat = labels.flatten()
    # for num in labels_flat:
    #   self.labels.append(num)
    precision = precision_score(labels_flat, pred_flat, zero_division=0, average="weighted")
    recall = recall_score(labels_flat, pred_flat, zero_division=0, average="weighted")
    f1 = f1_score(labels_flat, pred_flat, zero_division=0, average="weighted")
    return precision, recall, f1

  def test(self, data_loader):
    self.model.eval()
    total_recall = 0
    total_precision = 0
    total_f1 = 0
    total_loss = 0

    # once = False;

    with torch.no_grad():
      for (reviews, labels) in tqdm(data_loader):
        # TODO(students): start

        reviews = reviews.to(self.device)
        labels = labels.to(self.device)

        outputs = self.model(reviews, labels=labels)

        batch_loss = outputs.loss
        batch_logits = outputs.logits

        # if not once:
        #   print(batch_logits)
        #   once = True

        # get precision, recall, f1.
        (batch_precision, batch_recall, batch_f1) = self.get_performance_metrics(batch_logits.detach().cpu().numpy(), labels.detach().cpu().numpy())
        total_precision += batch_precision
        total_recall += batch_recall
        total_f1 += batch_f1
        total_loss += batch_loss

        # TODO(students): end
    
    precision = total_precision/len(data_loader)
    recall = total_recall/len(data_loader)
    f1 = total_f1/len(data_loader)
    loss = total_loss/len(data_loader)

    return precision, recall, f1, loss

  def execute(self):
    test_dataset = DatasetLoader(self.test_data, self.tokenizer)
    test_data_loader = test_dataset.get_data_loaders(self.batch_size)

    test_precision, test_recall, test_f1, test_loss = self.test(test_data_loader)

    print()
    print(f'test_loss: {test_loss:.4f} test_precision: {test_precision:.4f} test_recall: {test_recall:.4f} test_f1: {test_f1:.4f}')

Testing the embeddings

With Fixed Test Data

In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = fixed_test_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_frozen_embeddings'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 52/52 [00:13<00:00,  3.90it/s]


test_loss: 1.0601 test_precision: 0.4635 test_recall: 0.4918 test_f1: 0.4594


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = fixed_test_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_top_2_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 52/52 [00:13<00:00,  3.88it/s]


test_loss: 1.0297 test_precision: 0.4362 test_recall: 0.4361 test_f1: 0.3668


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = fixed_test_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_top_4_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 52/52 [00:13<00:00,  3.87it/s]


test_loss: 1.0231 test_precision: 0.4227 test_recall: 0.4798 test_f1: 0.4340


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = fixed_test_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_all_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 52/52 [00:13<00:00,  3.89it/s]



test_loss: 1.0160 test_precision: 0.4448 test_recall: 0.4787 test_f1: 0.4369


With Random Test Data

In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = random_test_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_frozen_embeddings'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 37/37 [00:09<00:00,  3.95it/s]


test_loss: 0.9057 test_precision: 0.5494 test_recall: 0.5771 test_f1: 0.5451


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = random_test_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_top_2_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 37/37 [00:09<00:00,  3.93it/s]


test_loss: 0.9095 test_precision: 0.5224 test_recall: 0.5327 test_f1: 0.5040


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = random_test_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_top_4_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 37/37 [00:09<00:00,  3.93it/s]


test_loss: 0.8854 test_precision: 0.5327 test_recall: 0.5738 test_f1: 0.5253


In [ ]:
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = random_test_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_all_training'
tester = Tester(options)
tester.execute()

Processing data..


100%|██████████| 37/37 [00:09<00:00,  3.94it/s]


test_loss: 0.8844 test_precision: 0.5474 test_recall: 0.5963 test_f1: 0.5576


Testing for Analysis and Discussion

In [ ]:
# Testing for Analysis and Discussion
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = fixed_test_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_all_training'
tester = Tester(options)
tester.execute()

arr = []
pred = []
actual = []
for i in range(len(tester.preds)):
  if(tester.preds[i] != tester.labels[i]):
    arr.append(i+2)
    pred.append([i+2, tester.preds[i]])
    actual.append([i+2, tester.labels[i]])

# print(arr)
print(pred)
print(actual)
# print(len(arr) == len(pred) and len(pred) == len(actual))

# print(tester.labels)

Processing data..


100%|██████████| 52/52 [00:13<00:00,  3.87it/s]


test_loss: 1.0161 test_precision: 0.4430 test_recall: 0.4792 test_f1: 0.4399
[[2, 2], [3, 2], [5, 2], [7, 2], [8, 2], [9, 2], [10, 2], [11, 2], [12, 2], [13, 2], [14, 1], [15, 2], [17, 2], [18, 2], [19, 2], [20, 2], [22, 2], [23, 2], [25, 2], [26, 2], [28, 2], [29, 2], [31, 2], [34, 2], [35, 2], [36, 2], [41, 2], [42, 1], [43, 1], [46, 1], [47, 2], [49, 2], [51, 2], [52, 2], [53, 1], [56, 2], [58, 2], [59, 1], [61, 2], [65, 2], [66, 2], [67, 1], [70, 1], [71, 2], [72, 1], [73, 1], [74, 1], [82, 2], [83, 2], [84, 1], [86, 2], [87, 2], [89, 1], [94, 1], [103, 1], [104, 2], [105, 1], [106, 2], [108, 2], [111, 2], [112, 1], [121, 2], [123, 1], [124, 2], [125, 2], [129, 2], [131, 1], [134, 2], [135, 1], [136, 2], [137, 2], [139, 1], [140, 2], [141, 1], [142, 2], [143, 2], [145, 2], [148, 2], [149, 2], [151, 2], [152, 1], [153, 2], [154, 2], [155, 1], [160, 2], [161, 1], [163, 2], [164, 1], [166, 1], [168, 2], [170, 2], [171, 2], [172, 1], [174, 1], [176, 2], [178, 2], [179, 2], [180, 1], [

In [ ]:
fixed_test2_data = load_dataset("data/fixed_test4.csv")

In [ ]:
# Testing for Analysis and Discussion
options = {}
options['batch_size'] = BATCH_SIZE
options['device'] = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
options['test_data'] = fixed_test2_data
options['save_path'] = DISTILBERT_SAVE_PATH + '_all_training'
tester = Tester(options)
tester.execute()

print(len(tester.preds) == len(tester.labels))
print(tester.preds)
print(tester.labels)
# print(len(arr) == len(pred) and len(pred) == len(actual))

# print(tester.labels)

Processing data..


100%|██████████| 1/1 [00:00<00:00, 16.75it/s]


test_loss: 0.9426 test_precision: 0.1111 test_recall: 0.3333 test_f1: 0.1667
True
[2, 2, 2]
[1, 1, 2]


# Method 3: Few-shot performance on a large language model (GPT-3)

In [ ]:
# from transformers import AutoTokenizer
# from transformers import AutoModelForCausalLM, AutoTokenizer
# import pandas as pd




In [ ]:


# # Load the dataset
# train = pd.read_csv('data/train.csv')

# # Extract the necessary columns
# texts = train['text'].tolist()
# sentiments = train['sentiment'].tolist()

# # Convert the sentiment labels to numerical values
# sentiments = [0 if s == 'negative' else 1 if s == 'neutral' else 2 for s in sentiments]

# #Convert the numpy arrays to tensors, create a tensor dataset objects and then torch dataloaders for easy batching
# X_train = torch.tensor(texts).float()
# Y_train = torch.tensor(sentiments).long()
# # X_test = torch.tensor(X_test).float()
# # Y_test = torch.tensor(Y_test).long()

# train_tensorDataset = torch.utils.data.TensorDataset(X_train,Y_train)
# # test_tensorDataset = torch.utils.data.TensorDataset(X_test,Y_test)

# train_loader = torch.utils.data.DataLoader(dataset=train_tensorDataset, batch_size=16, shuffle=True)
# # test_loader = torch.utils.data.DataLoader(dataset=test_tensorDataset, batch_size=batch_size, shuffle=False)


KeyError: ignored

In [ ]:
import pandas as pd
import requests
import math

In [ ]:
#train_data = pd.read_csv('data/train.csv')
test_data = pd.read_csv('data/fixed_test.csv')
test_dataset = test_data['DOCUMENT'].to_list()
labels = test_data['TRUE_SENTIMENT'].to_list()
#train_data_loader = train_dataset.get_data_loaders(self.batch_size)
#val_dataset = DatasetLoader(self.val_data, self.tokenizer)
#val_data_loader = val_dataset.get_data_loaders(self.batch_size)

In [ ]:
hf_key = "Bearer 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" # Andrew's Token
#hf_key = "Bearer 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"
API_URL = "https://api-inference.huggingface.co/models/google/flan-t5-xxl"
headers = {"Authorization": hf_key}

In [ ]:
def query(payload):
	response = requests.post(API_URL, headers=headers, json=payload)
	return response.json()

In [ ]:
def run_flan_t5_xxl_prompting(df, key):
  #print(key)
  #print(df)
  predictions = []
  i = 0
  for idx, row in df.iterrows():
    print(row[key])
    predictions.append(query({"inputs": row[key]})[0]['generated_text'])
    print(i)
    i+=1
  df[f"{key}_predictions"] = predictions
  return df

In [ ]:
df = test_data
print(type(list(df['DOCUMENT'])))


<class 'list'>


In [ ]:
"Sentimental Analysis: Sample input: 'I am good' Sample output: 'positive'"
#Fewshot sample
"Description of Task: Sentimental Analysis with outputs, pos,neutral,neg, examples of task: Sample input: 'I am good' Sample output: 'positive'"

In [ ]:



# prompt = "Description of Task: Sentiment Analysis with outputs, positive,neutral,negative: \n\ninput: \"{I am a good person}\"\noutput: positive"
# examples = [
#     ("I love this product.", "positive"),
#     ("This product is terrible.", "negative"),
#     ("This product is just okay.", "neutral")
# ]
# for ex in examples:
#     prompt += "\n\"{}\" => {}".format(ex[0], ex[1])
#     continue
  
# print(prompt)
# output = query({"inputs": f'{prompt}'})
# print(output)

Description of Task: Sentiment Analysis with outputs, positive,neutral,negative: 

input: "{I am a good person}"
output: positive
"I love this product." => positive
"This product is terrible." => negative
"This product is just okay." => neutral
[{'generated_text': 'neutral'}]


In [ ]:
def give_initial_str():
    prompt_str = "Read the following input sentence and rate whether it is positive, neutral, or negative.\n\n"
    prompt_str += "Input Sentence: The World is a bright place. I love the flowers and environment\n"
    prompt_str += "Rating: Positive \n\n"
    prompt_str += "Input Sentence: That movie was just okay. It was average and something I wouldn't watch again.\n"
    prompt_str += "Rating: Neutral\n\n"
    prompt_str += "Input Sentence: That game was terrible. It was the most awful game I've played in a while\n"
    prompt_str += "Rating: Negative\n\n\n"
    prompt_str += "Now rate this sentence\n\n"
    return prompt_str


def predict_sentiment(text):
    size = 1000-459
    multiple = math.ceil(len(text)/size)
    average = 0
    for i in range(multiple):
      prompt_str = give_initial_str()
      prompt_str += f"Input Sentence: {text[:size]}"
      text = text[size:]
      # prompt_str += f"Input Sentence: {text}"
      prompt_str += "Rating:"
      output = query({"inputs": prompt_str})[0]['generated_text']
      output = output.lower()
      average += 0 if output == 'negative' else 1 if output == 'neutral' else 2
      
    average = average/multiple
    average = 'negative' if average <=0.5 else 'neutral' if average<=1.5 else 'positive'
    return average
    #return output_text.strip()


In [ ]:
# text = "But could it be that U.S. Sen. Ted Cruz of Texas does n't get personally grilled by constituents very often ? "
# sentiment = predict_sentiment("Description of Task: Sentimental Analysis with outputs, positive,neutral,negative, examples of task: input: 'I am good' Sample output: 'positive'")
# sentiment = predict_sentiment(text)
# #sentiment = predict_sentiment(text)
# print(sentiment) # prints 'positive'



Neutral


In [ ]:
# text = "This was an okay movie."
# #sentiment = predict_sentiment("Description of Task: Sentimental Analysis with outputs, pos,neutral,neg, examples of task: Sample input: 'I am good' Sample output: 'positive'")
# #sentiment = predict_sentiment(df['DOCUMENT'][100])
# sentiment = predict_sentiment(text)
# print(sentiment) # prints 'positive'



positive


In [ ]:
#df = run_flan_t5_xxl_prompting(df, 'prompt1')

In [ ]:
def run_flan_t5_xxl_prompting_modified(df, key, starting_index):
  #print(key)
  #print(df)
  predictions = []
  i = 0
  for idx, row in df.iterrows():
    if (i<starting_index):
      i+=1
      continue
    print('index: ', i)
    try:
      predictions.append(predict_sentiment(row[key]))
    except:
      break
    i+=1
  return predictions

In [ ]:
df_predictions = []

starting_index = 0
full_length = len(df['DOCUMENT'])

df_predictions = run_flan_t5_xxl_prompting_modified(df,'DOCUMENT', starting_index)
  




index:  0


In [ ]:
# Assuming you have a DataFrame called 'data' with your predictions
data = pd.DataFrame({'Prediction': df_predictions})
# print(data)
print(df_predictions)

['negative', 'neutral', 'neutral', 'neutral', 'positive', 'neutral', 'neutral', 'neutral', 'negative', 'negative', 'neutral', 'neutral', 'negative', 'neutral', 'positive', 'neutral', 'neutral', 'neutral', 'neutral', 'neutral', 'positive', 'negative', 'neutral', 'negative', 'neutral', 'positive', 'negative', 'positive', 'negative', 'positive', 'neutral', 'negative', 'neutral', 'negative', 'negative', 'negative', 'neutral', 'negative', 'positive', 'neutral', 'neutral', 'negative', 'negative', 'negative', 'neutral', 'negative', 'negative', 'negative', 'positive', 'negative', 'neutral', 'negative', 'negative', 'neutral', 'positive', 'negative', 'negative', 'neutral', 'neutral', 'negative', 'neutral', 'negative', 'neutral', 'negative', 'neutral', 'negative', 'negative', 'negative', 'neutral', 'neutral', 'negative', 'neutral', 'negative', 'neutral', 'neutral', 'negative', 'positive', 'neutral', 'positive', 'neutral', 'negative', 'neutral', 'negative', 'negative', 'neutral', 'negative', 'posi

In [ ]:
# To create the file

#data.to_csv('few_shot_predictions.csv', index=False)
#data.to_csv('few_shot_predictions.txt', index=False)


In [ ]:
#When file is already created
# Read the existing CSV file into a DataFrame
df = pd.read_csv('few_shot_predictions.csv')

# Define the new data as a DataFrame

# Append the new data to the existing DataFrame
df = df.append(data, ignore_index=True)

# Write the updated DataFrame back to the CSV file
df.to_csv('few_shot_predictions.csv', index=False)
df.to_csv('few_shot_predictions.txt', index=False)


<ipython-input-16-cad8158a0f5e>:8: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(data, ignore_index=True)


In [ ]:
import pandas as pd
from sklearn.metrics import precision_score, recall_score, f1_score
df = pd.read_csv('data/few_shot_predictions.csv')
df['Sentiment'] = df['Prediction'].apply(lambda x: 0 if x == 'negative' else 1 if x == 'neutral' else 2)
df_true = pd.read_csv('data/fixed_test.csv')
df_true['TRUE_SENTIMENTS_CONVERTED'] = df_true['TRUE_SENTIMENT'].apply(lambda x: 0 if x == 'Negative' else 1 if x == 'Neutral' else 2)
ypred = df['Sentiment'].to_list()
y = df_true['TRUE_SENTIMENTS_CONVERTED'].to_list()

precision = precision_score(y, ypred, average='weighted')
recall_score =  recall_score(y, ypred, average='weighted')
f1_score = f1_score(y, ypred, average='weighted')

print('MODEL SCORES: ')
print('Precision: ', precision)
print('Recall: ', recall_score)
print('F1 Score: ', f1_score)

MODEL SCORES: 
Precision:  0.521093099211515
Recall:  0.3808948004836759
F1 Score:  0.38414043590828256


MODEL SCORES: 
1. Accuracy:  0.38089480048367597
2. Precision:  0.4264705882352941
3. Recall:  0.3625
4. F1 Score:  0.39189189189189183

NEW:
1. Precision:  0.521093099211515
2. Recall:  0.3808948004836759
3. F1 Score:  0.38414043590828256

In [ ]:
import pandas as pd

# Create a sample DataFrame

# Find rows where values in Column1 and Column2 are not the same
differing_rows = df[df_true['TRUE_SENTIMENTS_CONVERTED'] != df['Sentiment']]

# Print the differing rows and their indices
total = 0
for index, row in differing_rows.iterrows():
  total+=1
  print(f"Index: {index}")
print(total)
print('Accuracy: ', 1-total/len(df['Sentiment']))


Index: 0
Index: 2
Index: 5
Index: 11
Index: 12
Index: 18
Index: 19
Index: 20
Index: 22
Index: 23
Index: 27
Index: 28
Index: 29
Index: 30
Index: 31
Index: 32
Index: 33
Index: 35
Index: 36
Index: 37
Index: 40
Index: 41
Index: 42
Index: 43
Index: 44
Index: 45
Index: 46
Index: 47
Index: 51
Index: 52
Index: 53
Index: 54
Index: 55
Index: 57
Index: 58
Index: 59
Index: 60
Index: 61
Index: 63
Index: 65
Index: 66
Index: 67
Index: 68
Index: 70
Index: 71
Index: 72
Index: 73
Index: 74
Index: 75
Index: 77
Index: 79
Index: 80
Index: 82
Index: 83
Index: 85
Index: 87
Index: 88
Index: 89
Index: 90
Index: 92
Index: 93
Index: 94
Index: 95
Index: 96
Index: 97
Index: 98
Index: 99
Index: 100
Index: 101
Index: 102
Index: 103
Index: 104
Index: 105
Index: 107
Index: 108
Index: 109
Index: 110
Index: 111
Index: 112
Index: 114
Index: 115
Index: 119
Index: 121
Index: 122
Index: 123
Index: 124
Index: 125
Index: 128
Index: 136
Index: 137
Index: 138
Index: 140
Index: 143
Index: 144
Index: 145
Index: 146
Index: 147
Ind